### ISSUE TITLE
Simplenote users not able to log in with WordPress.com account on Mac, Windows, Linux

### ISSUE THREAD

**Issue by msilbers** [Automattic/simplenote-electron#1229]
Two reports so far for Linux (reproducible on Mac and Windows as well)
Ubuntu 18.10 with 4.18 kernel. Simplenote 1.5.0
Linux Mint and Manjaro. Both are the latest LTS version

I can reproduce this on the Electron Mac version as well, and there is a report for Windows, so it's all 3 Electron apps.

#### Steps to reproduce
1. Try to sign in via WordPress.com on Linux Simplenote app
2. Sign in, click Auth

#### What happened
Linux: "the authorization gets stuck when after clicking "Authorize" button"
Linux: "It goes berserk and issues thousands of requests to `auth`"
Mac Electron: for me, nothing happens at all

#### Simplenote version
1.5.0 (is not an issue in 1.4.1)

### **WORKAROUND**:
You can still log into the Simplenote app with your account details, using your email address and password used to originally sign up.

**Comment by gellenburg** [comment]
Having the same issue on the Windows 10 / Microsoft Store app version as well. This isn't limited to Linux. Can login just fine from Android. App is stuck on:

> Howdy! Simplenote would like to connect to all of your WordPress.com sites.

**Comment by liviopv** [comment]
I've been asking users to remove Simplenote from WordPress.com connected apps in https://wordpress.com/me/security/connected-applications to see if clearing that allows them to login. I believe I answered questions from the 2 Linux users Melissa talked about and another one with login issues in Windows.

Waiting on the user feedback.

**Comment by gellenburg** [comment]
Removing SimpleNote and re-adding it didn't help.

Still unable to login.

The debug console shows a bunch of messages though under the Network tab:

When I click on the Authorize button, SimpleNote calls the following URI:

https:// public-api.wordpress.com/oauth2/login/?client_id=99999&redirect_uri=https%3A%2F%2Fapp.simplenote.com%2Fwpcc&response_type=code&scope=global&state=app-6ef47a0237be4f999999&action=oauth2-login&redirect_to=https%3A%2F%2Fpublic-api.wordpress.com%2Foauth2%2Fauthorize%2F%3Fclient_id%3D99999%26redirect_uri%3Dhttps%253A%252F%252Fapp.simplenote.com%252Fwpcc%26response_type%3Dcode%26scope%3Dglobal%26state%3Dapp-6ef47a0237be4f999999%26jetpack-code%26jetpack-user-id%3D0%26action%3Doauth2-login&blog_id=0&_wpnonce=9999999999

Which responds with an HTTP/302 Redirect

location:
https:// app.simplenote.com/wpcc?code=XXXXXXXXX&state=app-6ef47a0237be4f999999`

And sets various cookies related to my account ID and credentials.

SimpleNote then calls that redirected URI above:

`Request URL:
  https:// app.simplenote.com/wpcc?code=XXXXXXXXXX&state=app-6ef47a0237be4f999999\

Which receives an HTTP/302 Redirect to the following URI (callback):

`location:
simplenote://auth?state=app-6ef47a0237be4f999999&user=REDACTED&token=0a2a...7e70&wp_token=17QE...0PMi&new=false`

But SimpleNote can't complete the call-back!

It tries to request:

`Request URL:
simplenote://auth?state=app-6ef47a0237be4f999999&user=REDACTED&token=0a2a...7e70&wp_token=17QE...0PMi&new=false`

Which never completes!

![2019-03-13_17-10-11](https://user-images.githubusercontent.com/7026750/54315090-af0f7880-45b3-11e9-9050-80c5652eba01.png)

**Comment by msilbers** [comment]
> Still unable to login.

@gellenburg for now, are you able to log in with the email address and password originally used to set up your Simplenote account?

**Comment by liviopv** [comment]
#1875140-zen also was not able to login.

**Comment by liviopv** [comment]
I've been able to reproduce it on a Mac as well. It get stuck on:

![Screenshot](https://cld.wthms.co/xNEU2Y+)
Screenshot: https://cld.wthms.co/xNEU2Y

**Comment by gellenburg** [comment]
How? By typing in my email and password.

Note, the web version of simplenote works just fine. It's the electron version that's borked.


On March 13, 2019 8:09:55 PM EDT, msilbers <notifications@github.com> wrote:
>> Still unable to login.
>
>@gellenburg for now, are you able to log in with the email address and
>password originally used to set up your Simplenote account?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/Automattic/simplenote-electron/issues/1229#issuecomment-472654314

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by msilbers** [comment]
> How? By typing in my email and password.

@gellenburg Yes, exactly. If you are having any trouble with that, you can contact support here and we can help you out: https://simplenote.com/contact-us/

**Comment by gellenburg** [comment]
> > How? By typing in my email and password.
> 
> @gellenburg Yes, exactly. If you are having any trouble with that, you can contact support here and we can help you out: https://simplenote.com/contact-us/

I'm confused, what is Support going to be able to help me with when the *Electron version* of SimpleNote is hanging when I attempt to login with my WordPress account, yet I am able to login with my WordPress Account with the *web version* of SimpleNote just fine?

**Comment by edequalsawesome** [comment]
Bug also reported in #1907138-zen

**Comment by pauljacobson** [comment]
I have a further report of this on Windows 10 in 1970162-zen. Here's the user's report:

>I use Windows 10 and my internet connection is 500/500 fiber, not proxy or VPN.
>
>I have installed the latest windows version of SimpleNote and do the following:
>
>1. Start the app
>
>2. At the login I click “Log in with WordPress.com”
>
>3. In the Authorize Simplenote window I click “Approve”
>
>4. The button turns to “Authorizing…” and then nothing happens. See attached screenshot.
>
>The authorization and login worked fine on my Mac and my Android phone.

I've recommended that the user sign in with an email address and password as a workaround.

**Comment by mriyazuddin** [comment]
I have a user on Ubuntu 18.04 with snap, Debian package or AppImage and can't login with their WordPress.com account. When they click on Approve button it is stuck on Authorizing...

ZD-1980519

I have advised the user to clear browser cookies and try. Also asked to try logging in with their email address instead of username.

**Comment by msilbers** [comment]
Issue reported again for Windows app in #2006605-zen
Advised them to log in with email address associated with account instead

**Comment by mgalbritton** [comment]
Reported again for windows app in #2017672-zen. Advised to log in with email address instead.

**Comment by tarunvijwani** [comment]
Reported again for Ubuntu app in #2016677-zen

**Comment by franckadil** [comment]
Can Reproduce on All clients (MacOS / Windows / Ubuntu / Arch Linux) this occurs on version: 1.5.0

![image](https://user-images.githubusercontent.com/42721509/57739059-fd232400-76a9-11e9-9239-825501d3028a.png)

**Comment by ItsJonQ** [comment]
Noticing this as well for https://github.com/Automattic/simplenote-electron/releases/tag/v1.5.0
Mac OS Mojave 10.14.4

**Comment by NGRhodes** [comment]
Affecting me too. 
Ubuntu 18.04, Simplenote 1.50

**Comment by gemmagarner** [comment]
Another report [on Twitter](https://twitter.com/ajmontpetit/status/1138085435916390400). Windows 10 and Linux (elementary OS 5.0)

**Comment by arturogiunta** [comment]
Same Issue. Windows 10 (downloaded exe and store app) Simplenote 1.5.0

**Comment by zdenys** [comment]
Another report in 2202705-zen on Windows: `I'm trying to log in to the Windows version using a Wordpress.com account, but I get stuck when I press OK.`

**Comment by Robertght** [comment]
Another report in #2274510-zen

**Comment by iSteveyang** [comment]
When i use simplenote without login with wordprees the cpu level is normal
![图片](https://user-images.githubusercontent.com/20753221/63234866-cbea5f00-c269-11e9-8c40-2bf953277ddb.png)
When i stuck on the cpu level is so weird
![图片](https://user-images.githubusercontent.com/20753221/63234961-3ef3d580-c26a-11e9-96fd-5f0c2b9af231.png)

**Comment by githubdh** [comment]
Is there an eta for a fix?

**Comment by dmsnell** [comment]
@codebykat after some auditing I stumbled upon some oddities.

Also see original PR in #764 

for one, we request the OAuth token with `request_type=code` but it looks like we are testing for the redirect URL to be the response from `request_type=token`. the `user`, `token`, and `state` parameters would come from this.

since we ask for `response_type=code` the API call responds with `simplenote://auth?code=SOMECODE`. that `SOMECODE` is supposed to then be used with a new `POST` call to `https://public-api.wordpress.com/oauth2/token` which then finally gives us our expected `token` inside the response as a JSON payload. _But we're not turning around and making that second request_ so the authentication never goes through.

on the other hand, _if_ we simply request `request_type=token` we get a response back with `set-cookie` HTTP headers containing the access token, not a redirect. I think that the `token` response is meant to set the cookie and have that be that. there aren't complicated redirect procedures involved. the `token` response type provides a time-limited access token, defaulting to two weeks.

---

it seems to me like we're probably just confused in this code. the web version and Electron versions _should_ use different mechanisms (web should have auto-expiring token probably, and Electron can store the code) but we're getting lost in the imperative code and interplay with the `BrowserWindow`.

a few thoughts:
 - the Electron docs are vague but we can create a new `session` for the `BrowserWindow` to start fresh with no cookies. this might help us also track things that are happening in the window vs. in the main app as it uses the default session for the application if none is provided
 - we might be spinning in a loop because in our `registerHttpProtocol( 'simplenote', … )` call we aren't doing anything with the request. I think we are allowing the existing request to continue along as-is which brings it right back to our registered handler. we should cancel the request and grab the `code` from the URL there in that handler.
 - did our OAuth process change? _no_, but the `token` request isn't providing a redirect! why not?

---

more manual testing. I used the "Persist Logs" function in the browser network tab and waited until I was ready to hit "Approve" on the "Hello!" screen. This one does the final redirect, and I found something.

When the `redirect_uri` is `https://simplenote.com` the response to the request is a `302 -> expected redirect URL with parameters` but when it's `simplenote://auth` the request returns a `200` with nothing but a cookie.

I'm suspicious that we might have an issue on the server when dealing with that protocol link. On the other hand, the test OAuth client you created uses `Native` when I think it should be `Web`. I created a clone of yours with this switch though and it didn't seem to matter.

At the heart of this issue is the `200 OK` instead of the `302 FOUND` response. If we find out why this its happening we'll probably discover why it's failing in this issue.

---

To summarize:
 - we need to talk about `request_type=token` vs `request_type=code`. I don't understand how this ever worked the way it does unless something changed in the OAuth server. There are tradeoffs to each approach: `request_type=token` is more secure but expires every two weeks; `request_type=code` is "permanent" but requires a second API call to get the real token.
 - something may be broken about `redirect_uri=simplenote://auth` because it's providing a `200` instead of a `302`
 - nothing forces us to use the `simplenote://` link since we can intercept all the redirects and do as we please. if we use `https://simplenote.com` then the redirect comes back right, though the app is still hunting for `searchParam`s when I believe it should be looking for `url.hash`. I ran into this [over three years ago](https://github.com/dmsnell/wrnc/blob/master/oauth-wrapper.jsx) and used the `url.hash` back then too, not the search parameters
 - I should have gone to sleep hours ago

**Comment by vietstone-ng** [comment]
I see this error now. I'm using the Windows desktop app, trying to signing in with Wordpress account. I stuck on this screen.
![Authorize Simplenote - Google Chrome 6_19_2020 12_01_38 PM](https://user-images.githubusercontent.com/744353/85098397-d5937900-b224-11ea-9eae-07f957d4f516.png)

**Comment by ElrikSettee** [comment]
hello from more than 2 years later and it still happens with electron on Mac Mojave. GRRR

**Comment by codebykat** [comment]
Hi @ElrikSettee, this is a quite old bug that was fixed in version 1.10, in 2019. If you are currently having trouble logging in can you please [open a new issue](https://github.com/Automattic/simplenote-electron/issues/new/choose) so we can help track down the problem? Thank you!
